File: go/serializers/dagr.go
package serializers

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"dagrv2/dagr"
	"dagrv2/documentgraphdirect"
	"dagrv2/eventgraphdirect"
	"dagrv2/messagegraphdirect"
	"dagrv2/stringsgraphdirect"
	"dagrv2/telemetrygraphdirect"

	"serializer-benchmark-go/model"
	modelv2 "serializer-benchmark-go/model/v2"
)

// dagrSer — Dagr ("Data Graph") via generated Go code.
//
// Schema: schemas/v2/dagr/schema.py → `dagr build` → go/gen/dagrv2 (its own module,
// wired with a `replace` in go.mod; layout "package-per-graph"). This row imports only the
// per-graph `<graph>direct` packages: each is self-contained (direct builder + lazy
// accessors); the `<graph>` packages (arenas + eager restore) serve the dagr-regular /
// dagr-frozen rows (dagr_regular.go, dagr_frozen.go). One DataGraph per suite type, all nodes packed,
// Telemetry.values marked `raw` (native-LE doubles, like protobuf's packed double).
//
// Serialize (timed): the generated **direct builder** (plain value structs → bytes, no
// arena). The value structs are this codec's native model, so — like protobuf's toProto —
// they are built from the suite values in Prepare (untimed); the timed call only runs
// `BuildAppend`. One `<Graph>Builder` per type is created in Prepare and reset by every
// Build; `BuildAppend` appends the record to a reused output buffer (Dagr writes
// back-to-front, so that one copy is inherent).
//
// Deserialize (timed): the generated **lazy reader** (`dagr.RootOffset` framing check +
// `New<Root>Accessor`, the body of `Open<Graph>Unchecked`; one recover per call so
// malformed bytes are an error, not a panic), then the
// owned suite value is materialized field by field. Arrays are walked with the
// generated `<F>Iter()` (O(n)), never `<F>At(i)` (O(i) per call).
//
// N>1 cells: the schema has no Batch_* wrapper (the harness frames N instances), so the
// frame is the suite's cross-language one (rust/src/run_v2.rs, C bench_serialize_cell):
// u32 LE count + (u32 LE len + record)×N. N=1 is the bare record.
//
// Per-type encode/decode functions are bound in Prepare (no type switch on the clock).
// Stream mode is adapted (Dagr's graph API is []byte-in / []byte-out).
//
// The same row type also serves the other three node layouts (dagr_layouts.go): only the
// Prepare-time binder differs, the timed path, framing and recover are shared.
type dagrSer struct {
	name string
	bind func(v any) (dagrCodec, error) // Prepare-time binder (owns any reusable builders)

	codec dagrCodec
	out   []byte // reused encode buffer; SerializeBytes returns a copy (as protobuf does)
}

// dagrCodec is one prepared cell.
type dagrCodec struct {
	enc func(dst []byte) []byte // timed: append the encoded value(s) to dst
	dec func(buf []byte) (any, error)
}

func newDagr() *dagrSer { return &dagrSer{name: "dagr-packed", bind: newDagrPackedBinder()} }

func (s *dagrSer) Name() string           { return s.name }
func (s *dagrSer) Version() string        { return dagrToolVersion() }
func (s *dagrSer) StreamMode() StreamMode { return StreamAdapted }
func (s *dagrSer) NativeKind() NativeKind { return NativeSchema }
func (s *dagrSer) Supports(n string) bool { return modelv2.IsV2TypeName(n) }

// ── version ────────────────────────────────────────────────────────────────

var (
	dagrVersionOnce sync.Once
	dagrVersion     string
)

// dagrToolVersion is the generator version recorded in the committed receipt
// schemas/v2/dagr/dagr.lock.json (provenance.tool_version "dagr 2026.9.2" → "2026.9.2") — the
// Go analogue of rust/build.rs's DAGR_VERSION. The generated code is vendored as a local
// module, so build info carries no version for it. go:embed cannot reach outside the Go
// module, so the receipt is read once at runtime from the repository root (the runner
// already resolves its run config the same way).
func dagrToolVersion() string {
	dagrVersionOnce.Do(func() {
		cwd, _ := os.Getwd()
		for dir := cwd; ; dir = filepath.Dir(dir) {
			raw, err := os.ReadFile(filepath.Join(dir, "schemas", "v2", "dagr", "dagr.lock.json"))
			if err == nil {
				var lock struct {
					Provenance struct {
						ToolVersion string `json:"tool_version"`
					} `json:"provenance"`
				}
				if json.Unmarshal(raw, &lock) == nil {
					dagrVersion = strings.TrimSpace(strings.TrimPrefix(lock.Provenance.ToolVersion, "dagr"))
				}
				return
			}
			if parent := filepath.Dir(dir); parent == dir {
				return
			}
		}
	})
	return dagrVersion
}

// ── prepare / timed path ───────────────────────────────────────────────────

func (s *dagrSer) Prepare(fx model.Fixture) error {
	c, err := s.bind(fx.Value)
	if err != nil {
		return fmt.Errorf("%s: %w", s.name, err)
	}
	s.codec = c
	s.out = s.out[:0]
	return nil
}

// newDagrPackedBinder binds the `dagr-packed` row: packed graphs, direct builder. One
// `<Graph>Builder` per type, created on first use and reset by every Build.
func newDagrPackedBinder() func(any) (dagrCodec, error) {
	var (
		msgB *messagegraphdirect.MessageGraphBuilder
		docB *documentgraphdirect.DocumentGraphBuilder
		telB *telemetrygraphdirect.TelemetryGraphBuilder
		strB *stringsgraphdirect.StringsGraphBuilder
		evB  *eventgraphdirect.EventGraphBuilder
	)
	return func(v any) (dagrCodec, error) {
		switch v.(type) {
		case modelv2.Message, []modelv2.Message:
			if msgB == nil {
				msgB = messagegraphdirect.NewMessageGraphBuilder()
			}
			return bindDagr(v, directMessage, msgB.BuildAppend, getMessage), nil
		case modelv2.Document, []modelv2.Document:
			if docB == nil {
				docB = documentgraphdirect.NewDocumentGraphBuilder()
			}
			return bindDagr(v, directDocument, docB.BuildAppend, getDocument), nil
		case modelv2.Telemetry, []modelv2.Telemetry:
			if telB == nil {
				telB = telemetrygraphdirect.NewTelemetryGraphBuilder()
			}
			return bindDagr(v, directTelemetry, telB.BuildAppend, getTelemetry), nil
		case modelv2.Strings, []modelv2.Strings:
			if strB == nil {
				strB = stringsgraphdirect.NewStringsGraphBuilder()
			}
			return bindDagr(v, directStrings, strB.BuildAppend, getStrings), nil
		case modelv2.Event, []modelv2.Event:
			if evB == nil {
				evB = eventgraphdirect.NewEventGraphBuilder()
			}
			return bindDagr(v, directEvent, evB.BuildAppend, getEvent), nil
		}
		return dagrCodec{}, fmt.Errorf("unsupported type %T", v)
	}
}

func (s *dagrSer) SerializeBytes(_ model.Fixture) ([]byte, error) {
	if s.codec.enc == nil {
		return nil, fmt.Errorf("%s: prepare() required before serialize", s.name)
	}
	s.out = s.codec.enc(s.out[:0])
	out := make([]byte, len(s.out))
	copy(out, s.out)
	return out, nil
}

func (s *dagrSer) DeserializeBytes(buf []byte) (v any, err error) {
	if s.codec.dec == nil {
		return nil, fmt.Errorf("%s: prepare() required before deserialize", s.name)
	}
	// The unchecked open trusts the bytes; a malformed buffer panics inside an
	// accessor — turn that into an error (one defer per call, not per field).
	defer func() {
		if r := recover(); r != nil {
			v, err = nil, dagr.ErrMalformed
		}
	}()
	return s.codec.dec(buf)
}

func (s *dagrSer) SerializeStream(fx model.Fixture, w io.Writer) (int, error) {
	return AdaptedSerializeStream(s, fx, w)
}

func (s *dagrSer) DeserializeStream(r io.Reader) (any, error) {
	return AdaptedDeserializeStream(s, r)
}

// bindDagr converts the prepared suite value(s) to direct value structs ONCE (untimed —
// the counterpart of protobuf's toProto in Prepare) and binds the single-record or the
// framed-batch codec, chosen from the value's shape.
func bindDagr[T, V any](
	sample any,
	conv func(T) V,
	put func(dst []byte, v V) []byte,
	get func(buf []byte) (T, error),
) dagrCodec {
	if x, single := sample.(T); single {
		v := conv(x)
		return dagrCodec{
			enc: func(dst []byte) []byte { return put(dst, v) },
			dec: func(buf []byte) (any, error) { return get(buf) },
		}
	}
	xs := sample.([]T)
	vs := make([]V, len(xs))
	for i := range xs {
		vs[i] = conv(xs[i])
	}
	return dagrCodec{enc: dagrFrameEnc(vs, put), dec: dagrFrameDec(get)}
}

// bindDagrArena is bindDagr for the arena serializers (regular / frozen roots): the
// arenas are built once in Prepare (untimed), and the timed encode stores each root into
// ONE reused Builder through the generated `AppendTo<Graph>` (reset, store, framing,
// append) — no Builder, dedup maps or result slice per record.
func bindDagrArena[T, R any](
	sample any,
	build func(T) R,
	appendTo func(b *dagr.Builder, dst []byte, root R) []byte,
	get func(buf []byte) (T, error),
) dagrCodec {
	b := dagr.NewBuilder(0)
	put := func(dst []byte, r R) []byte { return appendTo(b, dst, r) }
	if x, single := sample.(T); single {
		root := build(x)
		return dagrCodec{
			enc: func(dst []byte) []byte { return put(dst, root) },
			dec: func(buf []byte) (any, error) { return get(buf) },
		}
	}
	xs := sample.([]T)
	roots := make([]R, len(xs))
	for i := range xs {
		roots[i] = build(xs[i])
	}
	return dagrCodec{enc: dagrFrameEnc(roots, put), dec: dagrFrameDec(get)}
}

// dagrFrameEnc writes the suite's N>1 frame: u32 LE count + (u32 LE len + record)×N.
func dagrFrameEnc[V any](vs []V, put func(dst []byte, v V) []byte) func([]byte) []byte {
	return func(dst []byte) []byte {
		dst = binary.LittleEndian.AppendUint32(dst, uint32(len(vs)))
		for i := range vs {
			at := len(dst)
			dst = append(dst, 0, 0, 0, 0)
			dst = put(dst, vs[i])
			binary.LittleEndian.PutUint32(dst[at:], uint32(len(dst)-at-4))
		}
		return dst
	}
}

// dagrFrameDec reads the frame dagrFrameEnc writes, decoding each record with get.
func dagrFrameDec[T any](get func(buf []byte) (T, error)) func([]byte) (any, error) {
	return func(buf []byte) (any, error) {
		if len(buf) < 4 {
			return nil, fmt.Errorf("dagr: batch frame too short")
		}
		n := int(binary.LittleEndian.Uint32(buf))
		o := 4
		out := make([]T, n)
		for i := range out {
			if o+4 > len(buf) {
				return nil, fmt.Errorf("dagr: truncated batch frame")
			}
			l := int(binary.LittleEndian.Uint32(buf[o:]))
			o += 4
			if o+l > len(buf) {
				return nil, fmt.Errorf("dagr: truncated batch payload")
			}
			var err error
			if out[i], err = get(buf[o : o+l]); err != nil {
				return nil, err
			}
			o += l
		}
		return out, nil
	}
}

// ── suite value → direct value struct (Prepare, untimed; every field set) ───

func directMessage(m modelv2.Message) messagegraphdirect.Message {
	return messagegraphdirect.Message{
		FBool:    dagr.Some(m.FBool),
		FInt32:   dagr.Some(m.FInt32),
		FInt64:   dagr.Some(m.FInt64),
		FFloat64: dagr.Some(m.FFloat64),
		FString:  dagr.Some(m.FString),
		FBool2:   dagr.Some(m.FBool2),
		FInt32_2: dagr.Some(m.FInt32_2),
		FString2: dagr.Some(m.FString2),
	}
}

func directDocument(d modelv2.Document) documentgraphdirect.Document {
	items := make([]documentgraphdirect.DocumentItem, len(d.Items))
	for i, it := range d.Items {
		items[i] = documentgraphdirect.DocumentItem{
			Sku:        dagr.Some(it.SKU),
			Qty:        dagr.Some(it.Qty),
			PriceMinor: dagr.Some(it.PriceMinor),
		}
	}
	return documentgraphdirect.Document{
		ID:     dagr.Some(d.ID),
		Status: dagr.Some(d.Status),
		Meta: dagr.Some(documentgraphdirect.DocumentMeta{
			Region:  dagr.Some(d.Meta.Region),
			Version: dagr.Some(d.Meta.Version),
		}),
		Items: dagr.Some(items),
	}
}

func directTelemetry(t modelv2.Telemetry) telemetrygraphdirect.Telemetry {
	return telemetrygraphdirect.Telemetry{
		Source: dagr.Some(t.Source),
		Ts:     dagr.Some(t.TS),
		Tags:   dagr.Some(t.Tags),
		Values: dagr.Some(t.Values),
	}
}

func directStrings(st modelv2.Strings) stringsgraphdirect.Strings {
	return stringsgraphdirect.Strings{Items: dagr.Some(st.Items)}
}

func directEvent(e modelv2.Event) eventgraphdirect.Event {
	attrs := make([]eventgraphdirect.EventAttr, len(e.Attrs))
	for i, a := range e.Attrs {
		attrs[i] = eventgraphdirect.EventAttr{Key: dagr.Some(a.Key), Value: dagr.Some(a.Value)}
	}
	return eventgraphdirect.Event{
		EventID:    dagr.Some(e.EventID),
		EventType:  dagr.Some(e.EventType),
		OccurredAt: dagr.Some(e.OccurredAt),
		Producer:   dagr.Some(e.Producer),
		Attrs:      dagr.Some(attrs),
	}
}

// ── decode (lazy reader → owned suite value; strings are copied out) ──────

func getMessage(buf []byte) (modelv2.Message, error) {
	pos, err := dagr.RootOffset(buf)
	if err != nil {
		return modelv2.Message{}, err
	}
	a := messagegraphdirect.NewMessageAccessor(buf, pos)
	var m modelv2.Message
	m.FBool, _ = a.FBool()
	m.FInt32, _ = a.FInt32()
	m.FInt64, _ = a.FInt64()
	m.FFloat64, _ = a.FFloat64()
	m.FString, _ = a.FString()
	m.FBool2, _ = a.FBool2()
	m.FInt32_2, _ = a.FInt32_2()
	m.FString2, _ = a.FString2()
	return m, nil
}

func getDocument(buf []byte) (modelv2.Document, error) {
	pos, err := dagr.RootOffset(buf)
	if err != nil {
		return modelv2.Document{}, err
	}
	a := documentgraphdirect.NewDocumentAccessor(buf, pos)
	var d modelv2.Document
	d.ID, _ = a.ID()
	d.Status, _ = a.Status()
	if meta, ok := a.Meta(); ok {
		d.Meta.Region, _ = meta.Region()
		d.Meta.Version, _ = meta.Version()
	}
	it := a.ItemsIter()
	d.Items = make([]modelv2.DocumentItem, it.Len())
	for i := range d.Items {
		e, _ := it.Next()
		out := &d.Items[i]
		out.SKU, _ = e.Sku()
		out.Qty, _ = e.Qty()
		out.PriceMinor, _ = e.PriceMinor()
	}
	return d, nil
}

func getTelemetry(buf []byte) (modelv2.Telemetry, error) {
	pos, err := dagr.RootOffset(buf)
	if err != nil {
		return modelv2.Telemetry{}, err
	}
	a := telemetrygraphdirect.NewTelemetryAccessor(buf, pos)
	var t modelv2.Telemetry
	t.Source, _ = a.Source()
	t.TS, _ = a.Ts()
	tags := a.TagsIter()
	t.Tags = make([]string, tags.Len())
	for i := range t.Tags {
		t.Tags[i], _ = tags.Next()
	}
	vals := a.ValuesIter()
	t.Values = make([]float64, vals.Len())
	for i := range t.Values {
		t.Values[i], _ = vals.Next()
	}
	return t, nil
}

func getStrings(buf []byte) (modelv2.Strings, error) {
	pos, err := dagr.RootOffset(buf)
	if err != nil {
		return modelv2.Strings{}, err
	}
	a := stringsgraphdirect.NewStringsAccessor(buf, pos)
	it := a.ItemsIter()
	items := make([]string, it.Len())
	for i := range items {
		items[i], _ = it.Next()
	}
	return modelv2.Strings{Items: items}, nil
}

func getEvent(buf []byte) (modelv2.Event, error) {
	pos, err := dagr.RootOffset(buf)
	if err != nil {
		return modelv2.Event{}, err
	}
	a := eventgraphdirect.NewEventAccessor(buf, pos)
	var e modelv2.Event
	e.EventID, _ = a.EventID()
	e.EventType, _ = a.EventType()
	e.OccurredAt, _ = a.OccurredAt()
	e.Producer, _ = a.Producer()
	it := a.AttrsIter()
	e.Attrs = make([]modelv2.EventAttr, it.Len())
	for i := range e.Attrs {
		x, _ := it.Next()
		e.Attrs[i].Key, _ = x.Key()
		e.Attrs[i].Value, _ = x.Value()
	}
	return e, nil
}

File: javascript/src/serializers/dagr.js
/**
 * Dagr ("Data Graph") — schemas/v2/dagr/schema.py → `dagr build` → src/generated/dagr/*.ts,
 * compiled to src/generated/dagr_bundle.js by scripts/generate-dagr.sh (esbuild: the
 * generated TS uses extension-less imports and CI runs Node 20, so no native type stripping).
 *
 * The schema emits every suite type in four node layouts (spec/16-choosing-a-node-layout.md),
 * one row each: `dagr` (packed), `dagr-regular`, `dagr-frozen`, `dagr-frozen-packed`.
 *
 * packed / frozen+packed: serialize uses the generated **direct
 * builder** (`writeInto`: plain value objects → bytes, no arena) into one `Builder` reused
 * across calls (`reset()`). The value objects are the codec's native model, so — like the
 * protobuf libraries' messages — they are built in `prepare()` (untimed), every field set. Deserialize uses
 * the generated **lazy reader** (`<Type>Accessor.lazyRoot`) and materializes the domain
 * value inside the timed path (i64 fields are `bigint` on the Dagr side → Number).
 *
 * regular / frozen: the generator emits no direct builder for a non-packed root, so the
 * native model is the generated **arena** (`<Graph>_arena`, one arena per record, built in
 * `prepare()`); the timed path is only the arena serializer `<Graph>_serde.toBytes` (it
 * allocates its own `Builder` per call — the generated API offers no reuse). Deserialize is
 * the same lazy-reader → domain code for all four layouts (the accessor API is identical).
 *
 * Batch N>1: [u32 count][u32 len][record]… (same framing as the Rust harness).
 * Version: generator version from the committed receipt schemas/v2/dagr/dagr.lock.json.
 */
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import * as G from '../generated/dagr_bundle.js';

const { Builder } = G;

const __dirname = dirname(fileURLToPath(import.meta.url));

function dagrVersion() {
  try {
    const lock = JSON.parse(
      readFileSync(join(__dirname, '../../../schemas/v2/dagr/dagr.lock.json'), 'utf8'),
    );
    const m = /(\d+\.\d+\.\d+\S*)/.exec(String(lock.provenance?.tool_version ?? ''));
    return m ? m[1] : '';
  } catch {
    return '';
  }
}

const big = (v) => BigInt(v);
const num = (v) => (v === null ? 0 : Number(v));
const str = (v) => (v === null ? '' : v);

/* ---------- encode: direct-builder values (prepared, untimed) → bytes (timed) ---------- */

// One builder for the whole run: each encode resets it and stores the record at its far
// end; `recordBytes()` is a view of that record, valid until the next encode.
const B = new Builder(2 * 1024 * 1024, 4096);

// Domain → direct-builder value objects. These are the codec's native model (the
// counterpart of the protobuf libraries' message objects), so — like them — they are
// built in `prepare()`, outside the timer.
const converters = {
  message: (m) => ({
    f_bool: m.f_bool,
    f_int32: m.f_int32,
    f_int64: big(m.f_int64),
    f_float64: m.f_float64,
    f_string: m.f_string,
    f_bool_2: m.f_bool_2,
    f_int32_2: m.f_int32_2,
    f_string_2: m.f_string_2,
  }),
  document: (d) => ({
    id: d.id,
    status: d.status,
    meta: { region: d.meta.region, version: d.meta.version },
    items: d.items.map((it) => ({ sku: it.sku, qty: it.qty, price_minor: big(it.price_minor) })),
  }),
  telemetry: (t) => ({ source: t.source, ts: big(t.ts), tags: t.tags, values: t.values }),
  strings: (s) => ({ items: s.items }),
  event: (e) => ({
    event_id: e.event_id,
    event_type: e.event_type,
    occurred_at: big(e.occurred_at),
    producer: e.producer,
    attrs: e.attrs.map((a) => ({ key: a.key, value: a.value })),
  }),
};

const TYPES = {
  message: 'Message',
  document: 'Document',
  telemetry: 'Telemetry',
  strings: 'Strings',
  event: 'Event',
};

// Timed: the generated direct builder's writeInto (into the shared `B`).
const directWriters = (flavour) =>
  Object.fromEntries(
    Object.entries(TYPES).map(([id, T]) => {
      const Direct = G[`${T}${flavour}Direct`];
      return [id, (v) => Direct.writeInto(v, B)];
    }),
  );

// Domain → arena (regular / frozen: their native model; built in prepare(), untimed).
// One arena per record; the value is the root handle.
function arenaBuilders(flavour) {
  const A = (T) => G[`${T}${flavour}Arena`].Arena;
  return {
    message: (m) =>
      new (A('Message'))().newMessage(
        m.f_bool, m.f_int32, big(m.f_int64), m.f_float64,
        m.f_string, m.f_bool_2, m.f_int32_2, m.f_string_2,
      ),
    document: (d) => {
      const a = new (A('Document'))();
      const meta = a.newDocumentMeta(d.meta.region, d.meta.version);
      const items = d.items.map((it) => a.newDocumentItem(it.sku, it.qty, big(it.price_minor)));
      return a.newDocument(d.id, d.status, meta, items);
    },
    telemetry: (t) => new (A('Telemetry'))().newTelemetry(t.source, big(t.ts), t.tags, t.values),
    strings: (s) => new (A('Strings'))().newStrings(s.items),
    event: (e) => {
      const a = new (A('Event'))();
      const attrs = e.attrs.map((x) => a.newEventAttr(x.key, x.value));
      return a.newEvent(e.event_id, e.event_type, big(e.occurred_at), e.producer, attrs);
    },
  };
}

// Timed: the generated arena serializer (returns a fresh record).
const serdeWriters = (flavour) =>
  Object.fromEntries(
    Object.entries(TYPES).map(([id, T]) => {
      const Serde = G[`${T}${flavour}Serde`];
      return [id, (root) => Serde.toBytes(root)];
    }),
  );

/* ---------- decode: lazy reader → domain ---------- */

const makeDecoders = (flavour) => {
  const L = (T) => G[`${T}${flavour}Lazy`];
  return {
    message: (u8) => {
      const m = L('Message').MessageAccessor.lazyRoot(u8);
      return {
        f_bool: m.f_bool ?? false,
        f_int32: num(m.f_int32),
        f_int64: num(m.f_int64),
        f_float64: num(m.f_float64),
        f_string: str(m.f_string),
        f_bool_2: m.f_bool_2 ?? false,
        f_int32_2: num(m.f_int32_2),
        f_string_2: str(m.f_string_2),
      };
    },
    document: (u8) => {
      const d = L('Document').DocumentAccessor.lazyRoot(u8);
      const meta = d.meta;
      const src = d.items ?? [];
      const items = new Array(src.length);
      for (let i = 0; i < src.length; i++) {
        const it = src[i];
        items[i] = { sku: str(it.sku), qty: num(it.qty), price_minor: num(it.price_minor) };
      }
      return {
        id: str(d.id),
        status: num(d.status),
        meta: { region: str(meta?.region ?? null), version: num(meta?.version ?? null) },
        items,
      };
    },
    telemetry: (u8) => {
      const t = L('Telemetry').TelemetryAccessor.lazyRoot(u8);
      return { source: str(t.source), ts: num(t.ts), tags: t.tags ?? [], values: t.values ?? [] };
    },
    strings: (u8) => ({ items: L('Strings').StringsAccessor.lazyRoot(u8).items ?? [] }),
    event: (u8) => {
      const e = L('Event').EventAccessor.lazyRoot(u8);
      const src = e.attrs ?? [];
      const attrs = new Array(src.length);
      for (let i = 0; i < src.length; i++) {
        attrs[i] = { key: str(src[i].key), value: str(src[i].value) };
      }
      return {
        event_id: str(e.event_id),
        event_type: str(e.event_type),
        occurred_at: num(e.occurred_at),
        producer: str(e.producer),
        attrs,
      };
    },
  };
};

/* ---------- batch framing ---------- */

let scratch = new Uint8Array(1 << 16);

// Direct builder: stores into `B`; the record is a view valid until the next encode.
function encodeDirect(encode, value) {
  B.reset();
  encode(value);
  return B.recordBytes();
}

function encodeBatch(record, encode, values) {
  const n = values.length;
  let o = 4;
  for (let i = 0; i < n; i++) {
    const rec = record(encode, values[i]);
    const need = o + 4 + rec.length;
    if (need > scratch.length) {
      let cap = scratch.length * 2;
      while (cap < need) cap *= 2;
      const grown = new Uint8Array(cap);
      grown.set(scratch.subarray(0, o));
      scratch = grown;
    }
    scratch[o] = rec.length & 0xff;
    scratch[o + 1] = (rec.length >>> 8) & 0xff;
    scratch[o + 2] = (rec.length >>> 16) & 0xff;
    scratch[o + 3] = (rec.length >>> 24) & 0xff;
    scratch.set(rec, o + 4);
    o = need;
  }
  scratch[0] = n & 0xff;
  scratch[1] = (n >>> 8) & 0xff;
  scratch[2] = (n >>> 16) & 0xff;
  scratch[3] = (n >>> 24) & 0xff;
  return scratch.slice(0, o);
}

function decodeBatch(decode, u8) {
  const dv = new DataView(u8.buffer, u8.byteOffset, u8.byteLength);
  const n = dv.getUint32(0, true);
  const out = new Array(n);
  let o = 4;
  for (let i = 0; i < n; i++) {
    const len = dv.getUint32(o, true);
    o += 4;
    out[i] = decode(u8.subarray(o, o + len));
    o += len;
  }
  return out;
}

// Arena serializer: already returns a fresh, owned record.
const encodeSerde = (encode, root) => encode(root);

/**
 * One row per node layout. `flavour` is the graph-name infix of the bundle exports
 * ('' = packed, 'Regular', 'Frozen', 'FrozenPacked'); `direct` picks the direct builder
 * (packed-rooted graphs) over the arena + arena serializer.
 */
function makeDagrSer(name, flavour, direct) {
  const convs = direct ? converters : arenaBuilders(flavour);
  const writers = direct ? directWriters(flavour) : serdeWriters(flavour);
  const record = direct ? encodeDirect : encodeSerde;
  const decoders = makeDecoders(flavour);
  let encodeOne = null;
  let decodeOne = null;
  let isBatch = false;
  let prepared = null;   // native-model value(s) for the cell, built in prepare()
  return {
    name,
    version: dagrVersion(),
    category: 'schema',
    supports: (dataName) => Object.prototype.hasOwnProperty.call(writers, dataName),
    prepare(dataName, value) {
      const conv = convs[dataName];
      encodeOne = writers[dataName];
      decodeOne = decoders[dataName];
      if (!encodeOne) throw new Error(`${name}: no graph for ${dataName}`);
      isBatch = Array.isArray(value);
      prepared = isBatch ? value.map(conv) : conv(value);
    },
    serialize() {
      if (isBatch) return encodeBatch(record, encodeOne, prepared);
      // Direct: copy the record out of the reused builder; serde already owns its bytes.
      return direct ? encodeDirect(encodeOne, prepared).slice() : encodeOne(prepared);
    },
    deserialize(buf) {
      return isBatch ? decodeBatch(decodeOne, buf) : decodeOne(buf);
    },
  };
}

export const dagrSer = makeDagrSer('dagr', '', true);
export const dagrRegularSer = makeDagrSer('dagr-regular', 'Regular', false);
export const dagrFrozenSer = makeDagrSer('dagr-frozen', 'Frozen', false);
export const dagrFrozenPackedSer = makeDagrSer('dagr-frozen-packed', 'FrozenPacked', true);

export function dagrSerializers() {
  return [dagrSer, dagrRegularSer, dagrFrozenSer, dagrFrozenPackedSer];
}
